### PR TITLE
sql: add docs for rule_generate_column_substitute

### PR DESCRIPTION
**Pull request by toutdesuite**
### What is changed, added or deleted? (Required)

add docs for rule_generate_column_substitute

### Which TiDB version(s) do your changes apply to? (Required)

- [x] master (the latest development version)
- [x] v4.0 (TiDB 4.0 versions)
- [ ] v3.1 (TiDB 3.1 versions)
- [ ] v3.0 (TiDB 3.0 versions)
- [ ] v2.1 (TiDB 2.1 versions)

**If you select two or more versions from above**, to trigger the bot to cherry-pick this PR to your desired release version branch(es), you **must** add corresponding labels such as **needs-cherry-pick-4.0**, **needs-cherry-pick-3.1**, **needs-cherry-pick-3.0**, and **needs-cherry-pick-2.1**.

### What is the related PR or file link(s)?

- This PR is translated from: https://github.com/pingcap/docs-cn/pull/2940
- Other reference link(s):

### CODE REVIEW

File: generated-columns.md [hunk 1]
```diff
@@ -76,6 +76,57 @@ CREATE TABLE person (
 );
 ```
 
+## Index generation column replacement
```
Comment (ran-huang) [inline generated-columns.md:79]: `generation` or `generated`?

File: generated-columns.md [hunk 2]
```diff
@@ -76,6 +76,57 @@ CREATE TABLE person (
...
+
+When an expression that appears in a query has been stored as a generated column with an index, TiDB replaces the expression with the corresponding generated column. Therefore, you can use this index when generating the query plan.
```
Comment (ran-huang) [inline generated-columns.md:81]: `这样就可以在生成查询计划时考虑使用这个索引`
@LittleFall What is the subject of "考虑使用"? TiDB or the user?

File: generated-columns.md (hunk 1 above)
Comment (toutdesuite) [reply generated-columns.md]: @ran-huang Update. PTAL, thanks~

File: generated-columns.md (hunk 2 above)
Comment (toutdesuite) [reply generated-columns.md]: Update in 297c48450281f3dfadde985ab42b098abc33a46e~

File: generated-columns.md [hunk 3]
```diff
@@ -44,7 +44,7 @@ You can use the index on the stored generated column in order to speed up the fo
 SELECT name, id FROM person WHERE city = 'Beijing';
 ```
 
-If no data exists at path `$.city`, `JSON_EXTRACT` returns `NULL`. If you want to enforce a constraint that `city` must be `NOT NULL`, you can define the virtual column as follows:
+If no data exists at path `$.city`, `JSON_EXTRACT` returns `NULL`. If you want to enforce a constraint that `city` must be `NOT NULL`, you can define the `generated column` as follows:
```
Comment (wjhuang2016) [inline generated-columns.md:47]: virtual generated column

File: generated-columns.md [hunk 4]
```diff
@@ -56,7 +56,7 @@ CREATE TABLE person (
 );
 ```
 
-Both `INSERT` and `UPDATE` statements check virtual column definitions. Rows that do not pass validation return errors:
+Both `INSERT` and `UPDATE` statements check the `generated column` definition. Rows that do not pass validation return errors:
```
Comment (wjhuang2016) [inline generated-columns.md:59]: check the `virtual generated column` definition

**Comment by toutdesuite** [comment]
I' ll close this PR because changes have been made in https://github.com/pingcap/docs/pull/3051/files @wjhuang2016 @ran-huang FYI~

**Pull request closed without merging**